Slim/Control/Queries.pm

```diff
@@ -792,9 +794,31 @@ sub albumsQuery {
 			$request->addResultLoopIfValueDefined($loopname, $chunkCount, 'work_id', $c->{'tracks.work'});
 			$request->addResultLoopIfValueDefined($loopname, $chunkCount, 'work_name', $c->{'works.title'});
 			$request->addResultLoopIfValueDefined($loopname, $chunkCount, 'composer', $c->{'composer.name'});
-			$request->addResultLoopIfValueDefined($loopname, $chunkCount, 'grouping', $c->{'tracks.grouping'});
+			$request->addResultLoop($loopname, $chunkCount, 'grouping', $c->{'tracks.grouping'}||"");
+
+			# If we're dealing here with a Work rather than the full album, add relevant information into the album name and set favorites URL for the work. Moved to here
+			# from BrowseLibrary.pm so that it applies to all UIs.
+			my $albumTitle = $c->{'composer.name'} ? $c->{'composer.name'} . cstring($client, 'COLON') . ' ' : '';
+			if ( $c->{'tracks.work'} ) {
+				$albumTitle .= $c->{'works.title'} . ' (';
+				$albumTitle .= "$c->{'tracks.grouping'} " if $c->{'tracks.grouping'};
+				$albumTitle .= cstring($client,'FROM') . ' ';
+			}
+			$albumTitle .= $construct_title->();
+			$albumTitle .= ')' if $c->{'tracks.work'};
+
+			my $favoritesUrl = $c->{'tracks.work'}
```

Comment:
This again is not query code. Nothing else would define a favorites URL here. If you believe this must be unified, then create a helper somewhere outside. For all other cases this seems to be done in BrowseLibrary. Wouldn't this be possible for this case, too?

We really must make sure we keep some uniformity and consistency across the code. If something feels wrong, then it needs to be fixed for all of it, not just this new piece of code.

Reply:
Setting `favorites_url` in `BrowseLibrary` is flawed and can lead to inconsistency because it means that clients (such as Material) which don't use `BrowseLibrary` have to re-implement the logic, leading to code duplication and possible divergence.

`favorites_url`, though not directly a database column, is derived from the query results and is not something a client should mess with.

This has become necessary now because the simple construction of the favorites URL using `title` from `titles_loop` does not work when the "album" is actually an "album-work" (see previous reply about the title field itself).

I agree that we should fix this for all cases of `favorites_url`, not just for albums and works. If you agree, I will do that.

Reply:
But that again is the UI dev's choice: you don't want to use the protocol which is provided? Then you'll have to handle it with yourself. @CDrummond decided against using the browse menus, because he wanted to have more control and flexibility. That's absolutely fine, I have no problem with that! But then unfortunately he has to deal with many things himself. We can't implement every client's choice of options. We provide a ready-to-use menu, and we provide low-level queries. But we can't have everything in between.

Reply:
Or, we're consequent and provide those favourite URLs for all kinds of queries. But that would be a massive change of its own.

Reply:
I come back to the observation that a valid favorites_url is entirely defined by the query result. Unlike many other things, it's not something that a UI can mess with - all it can do (as Material has done up to now) is re-implement the logic from the browse menus. Otherwise, it wouldn't work as a URL.

So if the change needs to be implemented across all queries, for this PR to be accepted, then so be it.